common/src/main/java/tachyon/replay/ReplayCache.java

```diff
@@ -0,0 +1,192 @@
+ *
+ * Example usage:<br>
+ *
+ * <pre>
+ * <code>
+ * private final Cache<String, Long> cache = ReplayCache.create();
+ * ...
+ * public long myRpc(final boolean val1, final int val2, String rpcId) throws TachyonTException {
+ *   return mCache.run(rpcId, new ReplayCallable<Long>() {
+ *     @Override
+ *     public Long call() throws TachyonException {
+ *       return rpcWhichCanThrowTachyonException(val1, val2);
+ *     }
+ *   });
+ * }
+ * </code>
+ * </pre>
+ *
+ * @param <V> the type of value returned by this cache
+ */
+public final class ReplayCache<V> {
+  /** The maximum number of keys that can be cached before old keys are evicted. */
+  private static final long DEFAULT_MAX_SIZE = 10000;
+  /** The number of milliseconds a key may remain unaccessed in the cache before it is evicted. */
+  private static final long DEFAULT_EXPIRE_MS = 2000;
+
+  private final Cache<String, V> mCache;
+
+  /**
+   * Creates a cache with default maximum size and key expiration time.
+   *
+   * @return a cache with the default maximum size and expiration time
+   */
+  public static <V> ReplayCache<V> newReplayCache() {
+    return new ReplayCache<V>();
+  }
+
+  /**
+   * Creates a cache with the given maximum size and key expiration time.
+   *
+   * @param maxSize the maximum number of elements the cache may hold
+   * @param expireMs the amount of time to hold entries before they expire
+   * @return a cache with the specified maximum size and expiration time
+   */
+  public static <V> ReplayCache<V> newReplayCache(long maxSize, long expireMs) {
+    return new ReplayCache<V>(maxSize, expireMs);
+  }
+
+  /**
+   * Creates a cache with default maximum size and key expiration time.
+   */
+  private ReplayCache() {
+    this(DEFAULT_MAX_SIZE, DEFAULT_EXPIRE_MS);
+  }
+
+  /**
+   * Creates a cache with the given maximum size and key expiration time.
+   *
+   * @param maxSize the maximum number of elements the cache may hold
+   * @param expireMs the amount of time to hold entries before they expire
+   */
+  private ReplayCache(long maxSize, long expireMs) {
+    // TODO(andrew) Make it possible for users to configure this cache via
+    // CacheBuilder<Object, Object> from(String spec)
+    mCache = CacheBuilder.newBuilder().maximumSize(maxSize)
+        .expireAfterWrite(expireMs, TimeUnit.MILLISECONDS).<String, V>build();
+  }
+
+  /**
+   * The RPC handler method to be executed in {@link #run(String, ReplayCallable)}.
+   *
+   * @param <V> the return type of {@link #call()}
+   */
+  public interface ReplayCallable<V> {
+    V call() throws TachyonException;
+  }
+
+  /**
+   * Same with {@link ReplayCallable} except that this handler method throws
+   * {@link IOException} and is to be executed in {@link #run(String,
+   * ReplayCallableThrowsIOException}.
+   *
+   * @param <V> the return type of {@link #call()}
+   */
+  public interface ReplayCallableThrowsIOException<V> {
```

Comment:
The issue is that we need to specially handle `IOException` to transform it into a `ThriftIOException`, but we can't do that if the exception type is generic. It's basically the same situation as https://github.com/amplab/tachyon/pull/1801, where we have `RpcCallableThrowsTachyonTException`